resources/mirror.rb

```diff
@@ -87,19 +86,17 @@
 
   converge_if_changed do
     # if the mirror already exists, edit the configuration, otherwise create it
-    execute "Editing mirror - #{new_resource.mirror_name}" do
-      command "aptly mirror edit#{with_installer(new_resource.with_installer)}#{with_udebs(new_resource.with_udebs)}#{architectures(new_resource.architectures)}#{filter(new_resource.filter)}#{filter_with_deps(new_resource.filter_with_deps)}#{dep_follow_all_variants(new_resource.dep_follow_all_variants)}#{dep_follow_recommends(new_resource.dep_follow_recommends)}#{dep_follow_source(new_resource.dep_follow_source)}#{dep_follow_suggests(new_resource.dep_follow_suggests)}#{dep_verbose_resolve(new_resource.dep_verbose_resolve)} #{new_resource.mirror_name}"
-      user node['aptly']['user']
-      group node['aptly']['group']
-      environment aptly_env
-      only_if %(aptly mirror -raw list | grep ^#{new_resource.mirror_name}$)
-    end
+    mirror_command = if mirror_exists?(new_resource.mirror_name)
+                       "aptly mirror edit#{with_installer(new_resource.with_installer)}#{with_udebs(new_resource.with_udebs)}#{architectures(new_resource.architectures)}#{filter(new_resource.filter)}#{filter_with_deps(new_resource.filter_with_deps)}#{dep_follow_all_variants(new_resource.dep_follow_all_variants)}#{dep_follow_recommends(new_resource.dep_follow_recommends)}#{dep_follow_source(new_resource.dep_follow_source)}#{dep_follow_suggests(new_resource.dep_follow_suggests)}#{dep_verbose_resolve(new_resource.dep_verbose_resolve)}#{ignore_signatures(new_resource.ignore_signatures)} #{new_resource.mirror_name}"
+                     else
+                       "aptly mirror create#{with_installer(new_resource.with_installer)}#{with_udebs(new_resource.with_udebs)}#{architectures(new_resource.architectures)}#{filter(new_resource.filter)}#{filter_with_deps(new_resource.filter_with_deps)}#{dep_follow_all_variants(new_resource.dep_follow_all_variants)}#{dep_follow_recommends(new_resource.dep_follow_recommends)}#{dep_follow_source(new_resource.dep_follow_source)}#{dep_follow_suggests(new_resource.dep_follow_suggests)}#{dep_verbose_resolve(new_resource.dep_verbose_resolve)}#{ignore_signatures(new_resource.ignore_signatures)} #{new_resource.mirror_name} #{new_resource.uri} #{new_resource.distribution} #{new_resource.component}"
+                     end
```

Comment:
Seems like this should be a library helper method IMO.

Reply:
I'm a bit stuck on this one. I moved the code into a library function. The implementation looks like:

```ruby
    def mirror_command(res)
      if mirror_exists?(res.mirror_name)
        "aptly mirror edit#{with_installer(res.with_installer)}#{with_udebs(res.with_udebs)}#{architectures(res.architectures)}#{filter(res.filter)}#{filter_with_deps(res.filter_with_deps)}#{dep_follow_all_variants(res.dep_follow_all_variants)}#{dep_follow_recommends(res.dep_follow_recommends)}#{dep_follow_source(res.dep_follow_source)}#{dep_follow_suggests(res.dep_follow_suggests)}#{dep_verbose_resolve(res.dep_verbose_resolve)}#{ignore_signatures(res.ignore_signatures)} #{res.mirror_name}"
      else
        "aptly mirror create#{with_installer(res.with_installer)}#{with_udebs(res.with_udebs)}#{architectures(res.architectures)}#{filter(res.filter)}#{filter_with_deps(res.filter_with_deps)}#{dep_follow_all_variants(res.dep_follow_all_variants)}#{dep_follow_recommends(res.dep_follow_recommends)}#{dep_follow_source(res.dep_follow_source)}#{dep_follow_suggests(res.dep_follow_suggests)}#{dep_verbose_resolve(res.dep_verbose_resolve)}#{ignore_signatures(res.ignore_signatures)} #{res.mirror_name} #{res.uri} #{res.distribution} #{res.component}"
      end
    end
```
And the call in the mirror resource is `mirror_command(new_resource)`. It works fine when converging the node. However, the problem I have is with writing an effective chefspec test. Since the library function takes the `new_resource` object as an argument, the chefspec test needs to mock an equivalent resource object to pass into the function and this is where I'm stuck. I don't particularly want to attempt to re-implement the custom resource in a mock and I'm not sure how I can leverage the already existing custom resource object. Any thoughts on what I should do here?

Reply:
I believe this might be helpful:

https://github.com/sous-chefs/postgresql/blob/9e0efbfb0dbfb170349db4e8606e97a430348907/spec/libraries/helpers_spec.rb#L82-L97

Reply:
That's exactly what I needed. Thank you!